### DISCUSSION TITLE
Access input schema in the client

### DISCUSSION THREAD

**Discussion by solaldunckel** [trpc/trpc#2021] [Ideas]
Hi !

Would be awesome to have access to the input schema on the client for form validation.
It could be really useful if you pair it with packages such as **react-hook-form** that supports schema resolvers (zod, yup...).

Maybe it could be exposed through a utility function similar to that ?

```javascript
import { useForm } from 'react-hook-form';
import { inferMutationInput, trpc } from 'utils/trpc';
import { zodResolver } from '@hookform/resolvers/zod';

type FormData = inferMutationInput<'post.create'>;

const { register, handleSubmit } = useForm<FormData>({
  resolver: zodResolver(trpc.getInputSchema(['post.create'])),
});
```

**Comment by KATT** [comment]
See https://kitchen-sink.trpc.io/react-hook-form

**Reply by Axxxx0n** [reply]
Is there way to get it other way around? define on server and get it from query/mutation type?

**Reply by KATT** [reply]
It's a bad idea to import from the server but you can do a shared folder

**Reply by Max-DVS** [reply]
Quick question to avoid opening another thread on the same topic.
What happens if I create a file where I have the validation and in the same file I export the type with z.infer like: 

```
export const userValidationSchema = z.object({
  title: z.string().min(2),
  text: z.string().min(5),
});
export type User = z.infer<type of userValidationSchema>
```

Then use userValidationSchema on the route endpoint.
```
export const userRouter = createRouter()
  .mutation("createUser", {
    input: userValidation,
    async resolve({ input, ctx })
```

And in RHF
```
const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<User>({
    defaultValues: {
      title: "",
      text: "",
    },
    mode: "onBlur",
    resolver: zodResolver(userValidation),
  });
```

I would be breaking something with that approach ?

**Comment by kaminskypavel** [comment]
@KATT this feels suboptimal tbh, maybe it's just me but it makes sense that the server will "set the tone" and define its inputs which can be exported to the clients.

 It's a pretty common scenario where you have a single API serving multiple clients (web/expo/docs), whereas I haven't seen the other way around of having a single client who "demands" from several servers.

maybe it's a topic for a whole new discussion but it seems like there's a reasonable demand to develop a best practice that will unify tRPC + zod + prisma schemas.

**Comment by franco-roura** [comment]
You can create your own `InferMutationInput` type like this
```ts
type InferMutationInput<T extends UseTRPCMutationResult<any, any, any, any>> = NonNullable<
    T['variables']
  >
```

And use it like this 
![image](https://user-images.githubusercontent.com/58178352/200312580-78762b44-1e56-48cd-b4bf-e99414dd7bb8.png)

If that's the kind of interface you're looking for, then `useForm<InferMutationInput<typeof login>>()` should work

**Comment by KrallXZ** [comment]
I tried making and exporting validationSchema from the server and passing it to the `zodResolver`, but It's throwing error: "Error: You're trying to use @trpc/server in a non-server environment. This is not supported by default.". @KATT way is working but It seems weird to me to import validationSchema from client to server, not vice versa.

It would be great to have a method to get the validation schema from mutation.


But when it comes to type only @solaldunckel you can infer type like this:
```ts
type TEquipmentCreateInput = inferRouterInputs<Router>["create"];
```

where `Router` is type exported from your router, for example:
```ts
export type AppRouter = typeof appRouter;
```

**Comment by lucafaggianelli** [comment]
Hey there,

Just to share my solution, I ended up creating a shared folder in `src/schemas` and there I created a file for each model with all the schemas for that model, then I import the same schemas in both the tRPC routers and in the frontend.

Not optimal but at least it solves the problem and is less weird than importing something into the server from the frontend

**Comment by lucashfreitas** [comment]
@KATT not a huge priority, but I was thinking about this:

instead of creating a shared `package/schema` and share the zod object with the frontend/backend would it be possible for the trpc client to expose a getter where we could access the input/output zod object, since we already define it as a `.input` and `.output` on procedures.

Is this something that already exist or was discussed internally?

e.g

`trpc.schema.createUser.input` =>``// zod object instance i could pass to frontend form resolvers or other helpers``


```typescript

const form = useForm({
    resolver: zodResolver(trpc.schema.createUser.input),
    defaultValues: {
      ....
      ),
    },
  })
```

**Reply by Axedyson** [reply]
This

**Reply by KATT** [reply]
It works generally better now that zod has a native `toJsonSchema` and I've previously done examples how one could do an endpoint for this (ping me on discord and I can give it to you after the weekend)

However, not all zod schemas are serializable over the wire so it's hard to do for all forms

**Reply by LeulAria** [reply]
Im Looking for this aswell !
